File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int col0 = 1;
        int n = matrix.size();
        int m = matrix[0].size();

        // setting -1 in 1st row or coloumn ofr each 0 in ij
        for (int i = 0; i < n; i++){
            for (int j = 0; j < m; j++){
                if (matrix[i][j] == 0){
                    if (j == 0){
                        col0 = 0;
                    } else {
                        matrix[0][j] = 0;
                    }
                    matrix[i][0] = 0;
                }
            }
        }

        // turning -1 to 0s except 0rh row and 0th column
        for (int i = 1; i < n; i++){
            for (int j = 1; j < m; j++){
                if (matrix[i][0] == 0){
                    matrix[i][j] = 0;
                } else if (j > 0){
                    if (matrix[0][j] == 0)
                        matrix[i][j] = 0;
                } else if (j == 0 && col0 == 0){
                    matrix[i][j] = 0;
                }
            }
        }

        // now changing 0th row and 0th column
        if (col0 == 0){
            for (int i = 0; i < n; i++){
                matrix[i][0] = 0;
            }
        }
        if (matrix[0][0] == 0){
            for (int j = 0; j < m; j++){
                matrix[0][j] = 0;
            }
        }
    }
};

File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.java
class Solution {
    public void setZeroes(int[][] matrix) {
        boolean r00 = false;
        boolean c00 = false;

        if (matrix[0][0] == 0){
            r00 = true;
            c00 = true;
        }
        // checking for 0t row
        for (int i = 1; i < matrix[0].length; i++){
            if (matrix[0][i] == 0){
                r00 = true;
                break;
            }
        }

        // checking for 0th column
        for (int i = 1; i < matrix.length; i++){
            if (matrix[i][0] == 0){
                c00 = true;
                break;
            }
        }

        //checking remainning cells
        for (int i = 1; i < matrix.length; i++){
            for(int j = 1; j < matrix[0].length; j++){
                if (matrix[i][j] == 0){
                    matrix[0][j] = 0; //set 0 in 0th row
                    matrix[i][0] = 0;// set 0 in 0th column
                }
            }
        }

        //set the rest columns to zero
        for (int i = 1; i < matrix[0].length; i++){
            if (matrix[0][i] == 0){
                for (int j = 1; j < matrix.length; j++){
                    matrix[j][i] = 0;
                }
            }
        }

        //set the rest rows to zero
        for(int i = 1; i < matrix.length; i++){
            if (matrix[i][0] == 0){
                for (int j = 1; j < matrix[0].length; j++){
                    matrix[i][j] = 0;
                }
            }
        }

        //set 0th row to 0
        if (r00){
            for (int i = 0 ; i < matrix[0].length; i++){
                matrix[0][i] = 0;
            }
        }

        //set oth columns to 0
        if (c00){
            for(int i = 0; i < matrix.length; i++){
                matrix[i][0] = 0;
            }
        }
    }
}